toolkit/tools/scheduler/schedulerutils/priority.go

```diff
@@ -0,0 +1,240 @@
+	// the scheduler can still rely on the low-priority nodes to eventually provide the correct result.
+	logger.Log.Debugf("Setting priority for implicit provider nodes")
+	implicitNodes := subgraph.AllImplicitNodes()
+	satisfyingNodes := []*pkggraph.PkgNode{}
+	for _, implicitNode := range implicitNodes {
+		satisfyingNodesForImplicitNode, findErr := findMatchesForImplicit(implicitNode, pkgGraph)
+		if findErr != nil {
+			err = fmt.Errorf("failed to find match for implicit node (%s):\n%w", implicitNode.FriendlyName(), findErr)
+			return
+		}
+		if len(satisfyingNodesForImplicitNode) > 0 {
+			logger.Log.Debugf("For implicit (%s), we found (%v) that likely provides it.", implicitNode.FriendlyName(), satisfyingNodesForImplicitNode)
+			satisfyingNodes = append(satisfyingNodes, satisfyingNodesForImplicitNode...)
+		}
+	}
+
+	// For each real node we found, we want to prioritize it and all of its dependencies at high.
+	for _, realBuildNode := range satisfyingNodes {
+		nodePriorityMap.SetSubgraphPriority(pkgGraph, realBuildNode, HighNodePriority)
+	}
+
+	return
+}
+
+// IsElevatedPriority returns true if the given node is a medium or high priority node.
+func (s NodePriorityMap) IsElevatedPriority(node *pkggraph.PkgNode) bool {
+	pri := s.priorities[node]
+	return pri > LowNodePriority
+}
+
+// IncreaseNodePriority will increase the priority of the given node to the given priority. If the node already has a
+// higher priority, the priority will not be changed.
+func (s NodePriorityMap) IncreaseNodePriority(node *pkggraph.PkgNode, priority int) {
+	// Limit range of priority to MinPriority and MaxPriority
+	if priority < MinPriority {
+		priority = MinPriority
+	} else if priority > MaxPriority {
+		priority = MaxPriority
+	}
+
+	if s.priorities[node] < priority {
+		logger.Log.Tracef("Setting %s to priority %d", node.FriendlyName(), priority)
+		s.priorities[node] = priority
+	} else {
+		logger.Log.Tracef("Leaving %s at existing priority %d", node.FriendlyName(), s.priorities[node])
+	}
+}
+
+func (s NodePriorityMap) GetPriority(node *pkggraph.PkgNode) (priority int) {
+	return s.priorities[node]
+}
+
+// SetSubgraphPriority will increase the priority of the given node and all of its dependencies to the given priority.
+func (s NodePriorityMap) SetSubgraphPriority(pkgGraph *pkggraph.PkgGraph, node *pkggraph.PkgNode, priority int) {
+	search := traverse.BreadthFirst{}
+
+	// Walk the graph from the node and prioritize all its dependencies
+	search.Walk(pkgGraph, node, func(n graph.Node, d int) (stopSearch bool) {
+		pkgNode := n.(*pkggraph.PkgNode)
+		logger.Log.Warnf("Setting %s to priority %d", pkgNode.FriendlyName(), priority)
+		s.IncreaseNodePriority(pkgNode, priority)
+		return
+	})
+}
+
+// findMatchesForImplicit will try and find real packages that satisfies the implicit node. This is done through four methods:
+// 1. Find the best package node for the implicit node
```

Comment:
Just to make sure I grok this (please correct my understanding):

For a given implicit node (say, /bin/bash),
1. Look for an existing package (run node?) which provides said file 
2. if not found, search by rpmpath for  an exact match. How do we know an rpmpath for an implicit though? 
3. If both 1&2 fail, stop caring about finding a matching version for the provider nodes (the cached prebuilts?)

Reply:
2: See above, PMC will have given us a package when we ran the fetcher.